java/ql/src/experimental/Security/CWE/CWE-117/LogInjection.ql

```diff
@@ -0,0 +1,67 @@
+/**
+ * @name Log Injection
+ * @description Building log entries from user-controlled data is vulnerable to
+ *              insertion of forged log entries by a malicious user.
+ * @kind path-problem
+ * @problem.severity error
+ * @precision high
+ * @id java/log-injection
+ * @tags security
+ *       external/cwe/cwe-117
+ */
+
+import java
+import DataFlow::PathGraph
+import experimental.semmle.code.java.Logging
+import semmle.code.java.dataflow.FlowSources
+
+/**
+ * A taint-tracking configuration for tracking untrusted user input used in log entries.
+ */
+private class LogInjectionConfiguration extends TaintTracking::Configuration {
+  LogInjectionConfiguration() { this = "Log Injection" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) {
+    sink.asExpr() = any(LoggingCall c).getALogArgument()
+  }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    node.getType() instanceof BoxedType or node.getType() instanceof PrimitiveType
+  }
+
+  override predicate isSanitizerGuard(DataFlow::BarrierGuard guard) {
+    guard instanceof StrCheckSanitizerGuard
+  }
+}
+
+/**
+ * Models any regex or equality check as a sanitizer guard.
+ * Assumes any check on the taint to be a valid sanitizing check.
+ */
+private class StrCheckSanitizerGuard extends DataFlow::BarrierGuard {
+  StrCheckSanitizerGuard() {
+    exists(Method m |
+      m.getDeclaringType().hasQualifiedName("java.util.regex", "Pattern") and
+      m.hasName("matches")
+      or
+      m.getDeclaringType() instanceof TypeString and
+      m.hasName([
+          "startsWith", "regionMatches", "matches", "equals", "equalsIgnoreCase", "endsWith",
+          "contentEquals", "contains"
```

Comment:
I am a little bit sceptical whether most of these are really good sanitizers:

- `startsWith` / `endsWith`: Might only make sense when having an "allow" list for log message, but I don't think that this is really used in practice. It certainly cannot be used to detect line terminator characters.
- `equals` / `equalsIgnoreCase` / `contentEquals`: Would require that there is an "allow" list for log messages and the the log messages are known in advance, which is unlikely. It certainly cannot be used to detect line terminator characters in arbitrary log messages.
- `contains`: It is likely that this does not cover all bad input, see also https://github.com/github/codeql/pull/5099#discussion_r575690198. So even if this is used for sanitization, it is likely that the sanitization is incomplete.

Maybe it would be worth checking the difference when these methods are considered as guard, and when they are not (and whether their usage actually fully sanitizes the string).

Reply:
+1 @owen-mc Can you please run this PR twice once with the sanitizers and then without them?

Reply:
I agree that `startsWith`, `regionMatches` and `endsWith` aren't useful because it's hard to see how they could sanitize the string. You may as well leave `equals`, `equalsIgnoreCase` and `contentEquals` in, as they would sanitize, even if it seems unlikely that they would be used. I guess `contains` falls in the same category as `replace`, that it can be left out for now and added in later if a run on LGTM finds it occurring a lot.